packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts

```diff
@@ -189,13 +189,40 @@ interface EncodedFieldChange {
 }
 
 export class ModularEditBuilder extends ProgressiveEditBuilder<FieldChangeMap> {
+    private readonly fieldKinds: ReadonlyMap<FieldKindIdentifier, FieldKind>;
+
     constructor(
         family: ModularChangeFamily,
         deltaReceiver: (delta: Delta.Root) => void,
         anchors: AnchorSet,
     ) {
         super(family, deltaReceiver, anchors);
+        this.fieldKinds = family.fieldKinds;
+    }
+
+    submitChange(
+        path: UpPathWithFieldKinds | undefined,
+        field: FieldKey,
+        fieldKind: FieldKindIdentifier,
+        change: FieldChangeset,
+    ): void {
+        let nodeChange: FieldChangeMap = {};
+        nodeChange[field as string] = { fieldKind, change };
+
+        let remainingPath = path;
+        while (remainingPath !== undefined) {
+            const editor = getChangeHandler(this.fieldKinds, remainingPath.parentFieldKind).editor;
+            const fieldChange = editor.buildChildChange(remainingPath.parentIndex, nodeChange);
+            nodeChange = {};
+            nodeChange[field as string] = { fieldKind: remainingPath.parentFieldKind, change: brand(fieldChange) };
+            remainingPath = remainingPath.parent;
+        }
+
+        this.applyChange(nodeChange);
     }
+}
 
-    // TODO: Finish implementation
+export interface UpPathWithFieldKinds extends UpPath {
+    readonly parent: UpPathWithFieldKinds | undefined;
+    readonly parentFieldKind: FieldKindIdentifier;
 }
```

Comment:
This means we'll assign a field kind even if there is no parent, right? That seems right to me because the root field and detached field need some specific field kind/field editor. What I'm wondering is whether the name `parentFieldKind` still makes sense given all that. I don't feel very strongly about it but maybe something like `containingFieldKind` or just `fieldKind` would be more accurate. Although, one might argue it could be more confusing and that we should choose clarity over exactitude. Thoughts?

Reply:
`UpPath` always has a `parentField` and `parentIndex` even when `parent` is undefined, and there some documentation about this on `UpPath`.